agent.go

```diff
@@ -0,0 +1,346 @@
+	logger := loggerFromContext(ctx)
+
+	switch action.Name {
+	case "kubectl":
+		output, err := kubectlRunner(action.Input, a.Kubeconfig, workDir)
+		if err != nil {
+			return fmt.Sprintf("Error executing kubectl command: %v", err), err
+		}
+		return output, nil
+	case "cat":
+		output, err := bashRunner(action.Input, workDir, a.Kubeconfig)
+		if err != nil {
+			return fmt.Sprintf("Error executing cat command: %v", err), err
+		}
+		return output, nil
+	case "bash":
+		output, err := bashRunner(action.Input, workDir, a.Kubeconfig)
+		if err != nil {
+			return fmt.Sprintf("Error executing bash command: %v", err), err
+		}
+		return output, nil
+	default:
+		a.Trace(ctx, "system", fmt.Sprintf("Error: Tool %s not found", action.Name))
+		logger.Info("Unknown action: ", "action", action.Name)
+		return "", fmt.Errorf("unknown action: %s", action.Name)
+	}
+}
+
+func (a *Agent) AskLLM(ctx context.Context) (*ReActResponse, error) {
+	logger := loggerFromContext(ctx)
+	logger.Info("Asking LLM...")
+
+	data := Data{
+		Query:       a.Query,
+		PastQueries: a.PastQueries,
+		History:     a.History(),
+		Tools:       "kubectl, gcrane, bash",
+	}
+
+	prompt, err := a.generateFromTemplate(data)
+	if err != nil {
+		fmt.Println("Error generating from template:", err)
+		return nil, err
+	}
+
+	logger.Info("Asking LLM", "prompt", prompt)
+	WriteToFile(a.promptFilePath, prompt)
+	WriteToFile(a.promptFilePath, "\n========\n")
+
+	logger.Info("Thinking...", "prompt", prompt)
+
+	return a.ContentGenerator.GenerateContent(ctx, a.Model, prompt)
+}
+
+func sanitizeToolInput(input string) string {
+	return strings.TrimSpace(input)
+}
+
+func (a *Agent) Trace(ctx context.Context, role, content string) error {
+	logger := loggerFromContext(ctx)
+	logger.Info("Tracing...")
+	msg := Message{
+		Role:    role,
+		Content: content,
+	}
+	if role != "system" {
+		a.Messages = append(a.Messages, msg)
+	}
+	if a.tracePath == "" {
+		return nil
+	}
+	if err := WriteToFile(a.tracePath, fmt.Sprintf("%s: %s\n", role, content)); err != nil {
+		logger.Error("Error writing to trace file", "error", err)
+		// tracing is secondary to the main logic, so ignore the failures.
+	}
+	return nil
+}
+
+func (a *Agent) History() string {
+	var history strings.Builder
+	for _, msg := range a.Messages {
+		history.WriteString(fmt.Sprintf("%s: %s\n", msg.Role, msg.Content))
+	}
+	return history.String()
+}
+
+type ReActResponse struct {
+	Thought string  `json:"thought"`
+	Answer  string  `json:"answer,omitempty"`
+	Action  *Action `json:"action,omitempty"`
+}
+
+type Action struct {
+	Name   string `json:"name"`
+	Reason string `json:"reason"`
+	Input  string `json:"input"`
+}
+
+type Message struct {
+	Role    string `json:"role"`
+	Content string `json:"content"`
+}
+
+// WriteToFile appends the given content to a file.
+// If the file does not exist, it is created.
+// Returns an error if any operation fails.
+func WriteToFile(fileName string, content string) error {
```

Comment:
Nit: we should call this AppendToFile (and probably just refactor to a tracing layer)